protocol/src/lib.rs

```diff
@@ -246,6 +296,13 @@ pub struct SnapshotDetails {
 #[repr(u32)]
 #[derive(IntoPrimitive)]
 pub enum MessageVersion {
+    /// Changed `Write`, `WriteUnwritten`, and `ReadResponse` variants to have a
+    /// clean split between header and bulk data, to reduce `memcpy`
+    ///
+    /// Removed `#[repr(u16)]` and explicit variant numbering from `Message`,
+    /// because those are misleading; they're ignored during serialization.
```

Comment:
I don't quite understand this question.

`MessageVersion` must be the same on both sides, otherwise we'll fail negotiation (`HereIAm` / `YesItsMe`).  This hasn't changed in the PR.  After this PR, we'll need to update both Upstairs and Downstairs simultaneously.

This comment is explaining that I removed the `#[repr(u16)]` annotation from `enum Message` and explicit variant values from the first few variants, i.e. the trailing 3 in `VersionMismatch { version: u32 } = 3`.  Neither of these impacted serialization, which is determined by `#[derive(Serialize)]` (so it's just based on variant order, and ignores both `repr` and explicit values).

Reply:
It never worked like I expected, but it's not something that changed because of this PR.
I've filed https://github.com/oxidecomputer/crucible/issues/1210 to follow up with that another time.